### PR TITLE
Add features: erase mode and undo/redo function

### PR DESCRIPTION
**Pull request by xiasenhai**
**features：**
- erase mode
- undo/redo last draw

### CODE REVIEW

**Comment by xjiaoyang** [comment]
A better way of undo/redo implementation is on working

File: signature-pad/src/main/java/com/github/gcacace/signaturepad/views/SketchBoard.java [hunk 1]
```diff
@@ -0,0 +1,973 @@
...
+    private boolean validTrailNodes(List<TrailNode> nodes) {
+        boolean xChanged = false, yChanged = false;
+        TrailNode lastNode = null;
+        for (TrailNode node : nodes) {
+            if (lastNode == null) {
+                lastNode = node;
+                continue;
+            }
+            xChanged = node.eventX != lastNode.eventX;
+            yChanged = node.eventY != lastNode.eventY;
```
Comment (xjiaoyang) [inline signature-pad/src/main/java/com/github/gcacace/signaturepad/views/SketchBoard.java]: lastNode = node;

File: signature-pad/src/main/java/com/github/gcacace/signaturepad/views/SketchBoard.java [hunk 2]
```diff
@@ -0,0 +1,973 @@
...
+                            float eventX = node.getEventX();
+                            float eventY = node.getEventY();
+                            resetDirtyRect(eventX, eventY);
+                            timedPoint = new TimedPoint();
+                            timedPoint.x = eventX;
+                            timedPoint.y = eventY;
+                            timedPoint.timestamp = node.getTimestamp();
+                            addPoint(timedPoint);
+                        }
+                        mPoints.clear();
```
Comment (xjiaoyang) [inline signature-pad/src/main/java/com/github/gcacace/signaturepad/views/SketchBoard.java]: `drawerObject.data.clear();`

File: signature-pad/src/main/java/com/github/gcacace/signaturepad/views/SketchBoard.java [hunk 3]
```diff
@@ -0,0 +1,973 @@
...
+    // the trails we use to track the movement
+    private Trails trails;
+    /**
+     * state: REDO(2),UNDO(1),AVAILABLE(0)
+     */
+    private final AtomicInteger drawState = new AtomicInteger(0);
+    /**
+     * use to identify double click between two touches
+     */
+    private final AtomicInteger lastTouchCount = new AtomicInteger(0);
```
Comment (xjiaoyang) [inline signature-pad/src/main/java/com/github/gcacace/signaturepad/views/SketchBoard.java]: not need anymore!

**Comment by shobhitic** [comment]
This looks great, but why not squash all the commits into one?

**Comment by gcacace** [comment]
Hi, to accept the pull request, could you please squash all your commits?

**Comment by gcacace** [comment]
Hello,

please solve the conflicts and make sure it will work with the savedInstanceState bundle currently used to hold the bitmap in memory.

**Comment by 1uokun** [comment]
# How to undo?
```java
public ArrayList<Bitmap> mBitmapCaches =  new ArrayList<Bitmap>();

mSignaturePad.setOnSignedListener(new SignaturePad.OnSignedListener() {

     @Override
     public void onStartSigning() {
         //Event triggered when the pad is touched
     }

    @Override
    public void onSigned() {
        mBitmapCaches.add(mSignaturePad.getSignatureBitmap());
    }

    @Override
    public void onClear() {
        //Event triggered when the pad is cleared
    }
});

public void undo(){
        if(mBitmapCaches.size() < 2){
            mSignaturePad.clear();
            mBitmapCaches.clear();
        }else {
            mBitmapCaches.remove(mBitmapCaches.size()-1);
            mSignaturePad.setSignatureBitmap(mBitmapCaches.remove(mBitmapCaches.size()-1));
        }
}
```

**Pull request closed without merging**